cage.go

```diff
@@ -0,0 +1,163 @@
+	"strings"
+	"time"
+
+	"github.com/hf/nitrite"
+)
+
+var cageDialTimeout = 5 * time.Second
+
+// PCRs struct for attesting a cage connection against.
+type PCRs struct {
+	PCR0 string
+	PCR1 string
+	PCR2 string
+	PCR8 string
+}
+
+// Will return a http.Client that is connected to a specified cage hostname with a fully attested client.
+func (c *Client) CageClient(cageHostname string, expectedPCRs []PCRs) (*http.Client, error) {
+	c.expectedPCRs = expectedPCRs
+
+	caCertResponse, err := c.makeRequest(c.Config.EvervaultCagesCaURL, http.MethodGet, nil, "")
+	if err != nil {
+		return nil, err
+	}
+
+	cagesClient, err := c.cagesClient(cageHostname, caCertResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return cagesClient, nil
+}
+
+func (c *Client) cagesClient(cageHostname string, caCert []byte) (*http.Client, error) {
+	transport, err := c.cagesTransport(cageHostname, caCert)
+	if err != nil {
+		return nil, err
+	}
+
+	return &http.Client{Transport: transport}, nil
+}
+
+func (c *Client) cagesTransport(cageHostname string, caCert []byte) (*http.Transport, error) {
+	rootCAs, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("error getting system cert pool %w", err)
+	}
+
+	rootCAs.AppendCertsFromPEM(caCert)
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: false,
+		RootCAs:            rootCAs,
+		MinVersion:         tls.VersionTLS12,
+		ServerName:         cageHostname,
+	}
+
+	customDial := c.createDial(tlsConfig)
+
+	return &http.Transport{
+		DisableKeepAlives: true,
+		DialTLSContext:    customDial,
+	}, nil
+}
+
+// Attest that a given cert matches the expected PCRS.
+func attestCert(cert []byte, expectedPCRs []PCRs) (bool, error) {
+	// 1) extract the X509Certificate certificate from the bytes
+	certificate, err := x509.ParseCertificate(cert)
+	if err != nil {
+		return false, fmt.Errorf("unable to parse certificate %w", err)
+	}
+	// Extract the largest DNS name from the certificate
+	largestIndex := 0
+	for i := 1; i < len(certificate.DNSNames); i++ {
+		if len(certificate.DNSNames[i]) > len(certificate.DNSNames[largestIndex]) {
+			largestIndex = i
+		}
+	}
+
+	coseDNSValue := certificate.DNSNames[largestIndex]
+	// extract the COSE signature from the DNS name
+	coseSig := strings.Split(coseDNSValue, ".")[0]
+	// decode the hex encoded COSE signature
+	hexDecodedDNS, err := hex.DecodeString(coseSig)
+	if err != nil {
+		return false, fmt.Errorf("unable to decode certificate %w", err)
+	}
+
+	res, err := nitrite.Verify(hexDecodedDNS, nitrite.VerifyOptions{CurrentTime: time.Now()})
+	if err != nil {
+		return false, fmt.Errorf("unable to verify certificate %w", err)
+	}
+
+	verified := verifyPCRs(expectedPCRs, *res.Document)
+
+	return verified, nil
+}
+
+func verifyPCRs(expectedPCRs []PCRs, attestationDocument nitrite.Document) bool {
+	attestationPCRs := mapAttestationPCRs(attestationDocument)
+	for _, expectedPCR := range expectedPCRs {
+		isEqual := expectedPCR == attestationPCRs
+		return isEqual
+	}
```

Comment:
will this support partial matches? i.e. if I only specify a single PCR (e.g. PCR8) and leave the rest undefined